### PR TITLE
dap_adapters: add gdb

### PR DESCRIPTION
**Pull request by jansol**
Nothing special here, just launching gdb with `-i=dap`. Nicer to have it as a builtin type instead of a custom DAP though.

### CODE REVIEW

**Comment by RemcoSmitsDev** [comment]
Hey, yeah totally agree, thanks for adding this! 
Do you have a task example that we can share inside a specific Zed channel for GDB?

**Comment by Anthony-Eid** [comment]
Thank you for implementing this!

**Comment by Anthony-Eid** [comment]
> Hey, yeah totally agree, thanks for adding this! Do you have a task example that we can share inside a specific Zed channel for GDB?

I think it would be similar to lldb

```json
{
  "label": "LLDB Debug Zed",
  "program": "$ZED_WORKTREE_ROOT/target/debug/zed",
  "adapter": "gdb",
  "request": "launch"
  "initialize_args": {
      "stopAtBeginningOfMainSubprogram": true
   }
}
```

**Comment by RemcoSmitsDev** [comment]
Would it be similar to LLDB? I would like to add a default config for debugging Zed with GDB inside the `/.zed/debug.json` and inside the `initial-debug-tasks.json`.

```json
  {
    "label": "Debug Zed with LLDB",
    "adapter": "lldb",
    "program": "$ZED_WORKTREE_ROOT/target/debug/zed",
    "request": "launch",
    "cwd": "$ZED_WORKTREE_ROOT"
  }
```
GDB debug task example??
```json
  {
    "label": "Debug Zed with GDB",
    "adapter": "gdb",
    "program": "$ZED_WORKTREE_ROOT/target/debug/zed",
    "request": "launch",
    "cwd": "$ZED_WORKTREE_ROOT"
  }
```

**Comment by jansol** [comment]
I have this left over in the project where I tested last time:
```json
  {
    "label": "Launch in GDB",
    "adapter": "gdb",
    "session_type": "launch",
    "program": "$ZED_WORKTREE_ROOT/build/main-binary",
    "initialize_args": {
      "args": [],
      "cwd": "$ZED_WORKTREE_ROOT",
      "env": {},
      "stopAtBeginningOfMainSubprogram": true
    }
  }
```

But it seems the format has changed since then?

Also the `stopAtBeginningOfMainSubprogram` should really not be needed. It's just the only way I've gotten the debugger to actually open instead of exiting immediately.

File: crates/dap_adapters/src/gdb.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        &self,
+        _: &dyn DapDelegate,
+        _: &DebugAdapterConfig,
+        _: Option<PathBuf>,
+    ) -> Result<DebugAdapterBinary> {
+        unimplemented!("GDB cannot be installed by Zed (yet)")
+    }
+
+    fn request_args(&self, config: &DebugAdapterConfig) -> Value {
+        json!({"program": config.program})
```
Comment (RemcoSmitsDev) [inline crates/dap_adapters/src/gdb.rs]: Can you also pass in the **cwd** key? So other people don't have to think about adding it to their `initialize_args`.

**Review by RemcoSmitsDev — CHANGES_REQUESTED** [review]
Could you also add this to the `.zed/debug.json`? I wanted to push this change, but I cannot to your fork.

```json
  {
    "label": "Debug Zed with GDB",
    "adapter": "gdb",
    "program": "$ZED_WORKTREE_ROOT/target/debug/zed",
    "request": "launch",
    "cwd": "$ZED_WORKTREE_ROOT",
    "initialize_args": {
      "stopAtBeginningOfMainSubprogram": true
    }
  }
```

**Comment by RemcoSmitsDev** [comment]
I also saw that GDB support attach, we might want to implement that as well. Happy to help to make that work, but cannot get GDB to work on my MacBook.

**Comment by jansol** [comment]
Attaching to a process with GDB is indeed possible, but requires special permissions (SYS_PTRACE capability or root) so that should probably wait until we have some form of polkit integration for escalating privileges.

**Comment by RemcoSmitsDev** [comment]
Alright, thanks for the clarification and your contribution!

**Comment by Anthony-Eid** [comment]
We should add a cfg flag to disable gfb on m series macs because there’s no arm port available.

That way people can’t select it when it’s bot possible to use

**Comment by jansol** [comment]
GDB is apparently perfectly fine on aarch64, just not on macOS.

**Pull request merged**